Support materializing exchanges

arhimondr opened the pull request:
Attempt to #12387.

Depends on #12464. Don't review the first commit.

arhimondr commented:
I think many of the commits may have to be squashed before the merge. For the sake of reviewers convenience lets keep them for now.

wenleix commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/PhasedExecutionSchedule.java:
```diff
@@ -269,7 +268,7 @@ public Visitor(Collection<PlanFragment> fragments, DirectedGraph<PlanFragmentId,
         public Set<PlanFragmentId> visitExchange(ExchangeNode node, PlanFragmentId currentFragmentId)
         {
-            checkArgument(node.getScope() == LOCAL, "Only local exchanges are supported in the phased execution scheduler");
+            checkArgument(node.getScope().isLocal(), "Only local exchanges are supported in the phased execution scheduler");
```
Do we want to change this? (Or simply use `!node.getScope().isRemote()` ? )

arhimondr replied:
Let's keep `isLocal` method. It is easier to read than `!isRemote()`.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java:
```diff
@@ -57,7 +55,18 @@
+            return this == LOCAL;
+        }
+
+        public boolean isRemote()
```
What about using enum class with constructor ? (Similar to `BufferState` class)?

wenleix replied:
Update: I see, `JoinDistributionType` use this style, we can keep this for now since I don't have strong opinion :)

tdcmeehan replied:
I think passing into the constructor is the way to go here

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -863,7 +870,7 @@ public PlanWithProperties visitSemiJoin(SemiJoinNode node, PreferredProperties p
+                                partitionedExchange(idAllocator.getNextId(), getExchangeScope(source.getNode()), source.getNode(), createPartitioning(sourceSymbols), Optional.empty()),
                                 source.getProperties());
                         filteringSource = withDerivedProperties(
                                 partitionedExchange(idAllocator.getNextId(), REMOTE, filteringSource.getNode(), createPartitioning(filteringSourceSymbols), Optional.empty(), true),
```
I personally find it difficult to reason/follow why some remote partitioned exchange will call `getExchangeScope`, why some are hard coded to `REMOTE`. This might also make future code maintenance difficult. 

My understanding is we didn't enable it other than JOIN/AGGREGATE. One thought is to pass the "parentNode" to `getExchangeScope`.

Let's discuss in person once @rschlussel  also took a look

arhimondr replied:
https://github.com/prestodb/presto/pull/12469#issuecomment-478719893

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -863,7 +870,7 @@ public PlanWithProperties visitSemiJoin(SemiJoinNode node, PreferredProperties p
                     else {
                         source = withDerivedProperties(
-                                partitionedExchange(idAllocator.getNextId(), REMOTE, source.getNode(), createPartitioning(sourceSymbols), Optional.empty()),
+                                partitionedExchange(idAllocator.getNextId(), getExchangeScope(source.getNode()), source.getNode(), createPartitioning(sourceSymbols), Optional.empty()),
```
This is in SemiJoin, do we still want to materialize exchange ?

arhimondr replied:
Yes, it might be beneficial for retriability.

wenleix commented on presto-main/src/test/java/com/facebook/presto/execution/TestQueryManagerConfig.java:
```diff
@@ -21,6 +21,9 @@
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
+import static com.facebook.presto.execution.QueryManagerConfig.MaterializeExchangesStrategy.ALL;
```
`ALL` and `NONE` don't fit for static import since the name is too general :)

arhimondr replied:
I think the code like

`.setMaterializeExchangesStrategy(NONE)` is pretty readable.

Something like .set**MaterializeExchangesStrategy**(**MaterializeExchangesStrategy**.NONE) sounds like a tautology

wenleix commented on presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java:
```diff
@@ -394,4 +411,10 @@ public QueryManagerConfig setRequiredWorkersMaxWait(Duration requiredWorkersMaxW
         return this;
     }
+
+    public enum MaterializeExchangesStrategy
```
What about put this `MaterializeExchangesStrategy` into `FeaturesConfig.java` ? -- So it sits together with `JoinReorderingStrategy` and `JoinDistributionType` :)

arhimondr replied:
I think the idea was to keep enum in the same configuration class where it is used. All the enums are there used in the `SystemSessionProperties`. But if we want to group them by their usage in the `SystemSessionProperties` we should move them there then.

wenleix commented on presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java:
```diff
@@ -685,7 +685,19 @@ public Lock getExclusiveLock()
+                session,
+                plan,
+                true,
+                new PlanNodeIdAllocator()
```
I think the usual pattern in Presto code base is to make PlanNodeIdAllocator an interface, and have SimplePlanNodeIdAllocator and MockPlanNodeIdAllocator ? :)

wenleix replied:
Why we cannot just create a normal `PlanNodeIdAllocator` here ?

arhimondr replied:
> Why we cannot just create a normal PlanNodeIdAllocator here 

In theory you should always pass the same `PlanNodeIdAllocator` that was used in the logical planner. So the plan node ids don't clash.

But for the LocalQueryRunner specifically this is never supposed to be called (because for the LocalQueryRunner PlanFragmenter is pretty much no-op). 

So, instead of messing around many method signatures to properly pull down the `PlanNodeIdAllocator` i decided to pass something dummy, as it never gets called.

> and have SimplePlanNodeIdAllocator and MockPlanNodeIdAllocator

This implementation is more like `NotImplementedPlanNodeIdAllocator`. But since we need one only in one place i woudn't bother of creating a separate class for it.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -359,10 +393,20 @@ public PlanNode visitValues(ValuesNode node, RewriteContext<FragmentProperties>
         {
-            if (!exchange.getScope().isRemote()) {
-                return context.defaultRewrite(exchange, context.get());
+            switch (exchange.getScope()) {
```
It should be using switch when `REMOTE_MATERIALIZED` is introduced ;)

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -389,6 +433,242 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+            }
+
+            NewTableLayout insertLayout = metadata.getInsertLayout(session, tableHandle)
+                    // Currently it is assumed that the input data has to be always
```
I don't understand why these comments are here? -- Looks like this part of code is irrelevant to exchange ? :)

arhimondr replied:
Discussed offline

rschlussel commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -59,6 +60,7 @@
     public static final String DISTRIBUTED_INDEX_JOIN = "distributed_index_join";
     public static final String HASH_PARTITION_COUNT = "hash_partition_count";
     public static final String PARTITIONING_PROVIDER_CATALOG = "partitioning_provider_catalog";
+    public static final String MATERIALIZE_EXCHANGES_STRATEGY = "materialize_exchanges_strategy";
```
nit: `exchange_materialization_strategy` sounds better to me (it works as an English phrase).

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -15,6 +15,7 @@
 
 import com.facebook.presto.Session;
 import com.facebook.presto.connector.system.GlobalSystemConnector;
+import com.facebook.presto.execution.QueryManagerConfig.MaterializeExchangesStrategy;
```
The materialization choice doesn't depend on any of the complicated logic in add exchanges.  Instead of adding still more complexity to AddExchanges, could we have a simple optimizer that runs after add exchanges and converts REMOTE exchanges to REMOTE_MATERIALIZED exchanges.  That will also make it simpler to introduce costing to that choice later.

arhimondr replied:
This is a very good comment.

We had a similar discussion with @wenleix. Eventually we decided that it is not a good idea to split the AddExchanges logic in multiple places. The AddExchanges optimizer is very complex, and we don't want to endup with two different classes that partially duplicate the same logic.

Although currently the logic is straightforward, it might not be so once we support materialized exchanges for the cases when one of the join tables is already bucketed.

rschlussel replied:
Based on offline discussions, it seems like the bigger issue is that ability to materialize depends on using connector partitioning, which has to be set in AddExchanges so that nodes higher up in the tree can derive the correct partitioning properties.  For now we'll keep the materialization decision here, and we'll extract it when we need to.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java:
```diff
@@ -57,7 +55,18 @@
     public enum Scope
```
I don't think scope makes so much sense as a name anymore since it also indicates whether it's materialized, but I don't have a better suggestion, since `Type` is already taken.

wenleix replied:
`Channel` as I suggested in https://github.com/prestodb/presto/pull/12419#pullrequestreview-209896190 ;) 

Although I feel this would make a lot of changes. So maybe we can just keep the legacy `Scope` name here ....

rschlussel replied:
If we also rename the enum values as  you suggested, channel seems fine to me.  But as you said that would be a lot of changes (though not hard to do since intellij can do it for you).  for now I'm ok leaving "Scope" until the usage settles a bit.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java:
```diff
@@ -57,7 +55,18 @@
     {
         LOCAL,
-        REMOTE
+        REMOTE,
```
I wonder if this should also have a modifier now, e.g. `REMOTE_STREAMING`

wenleix replied:
`REMOTE_NETWORK` ? :)

arhimondr replied:
I think `REMOTE_STREAMING` is a better candidate, because both of them are in some sense over the network.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java:
```diff
@@ -98,15 +107,15 @@ public ExchangeNode(
                 "local exchanges do not support constant partition function arguments");
 
-        checkArgument(scope != REMOTE || type == REPARTITION || !partitioningScheme.isReplicateNullsAndAny(), "Only REPARTITION can replicate remotely");
+        checkArgument(!scope.isRemote() || type == REPARTITION || !partitioningScheme.isReplicateNullsAndAny(), "Only REPARTITION can replicate remotely");
```
also check that if it's remote_materialized that the type is REPARTITION

wenleix commented:
BTW: I think UI is currently broken when you materialize exchange....

wenleix commented:
TEST_SPECIFIC_MODULES=presto-hive is failing since
```
The job exceeded the maximum log length, and has been terminated.
```

Maybe consider split into `presto-hive` and `presto-hive-materialized-exchange`?

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -699,6 +700,11 @@ public FragmentProperties(PartitioningScheme partitioningScheme)
             return children;
         }
 
+        public Set<PlanFragmentId> getDependencies()
```
Now `children` and `dependencies` are interesting:
```
- `children` is `SubPlan`, and (when executes) are connected by remote exchange.
- `dependencies` is `PlanFragment`, and (when executes) are connected by TableWrite + TableScan.
```

This is not so easy to understand at first glance, since there are two types of "dependency" and they are tracked by two different classes. Can they be both tracked by `SubPlan` ? :)  I will also tries to read the later commits. 

BTW: Is there any reason to use `PlanFragmentId` instead of `PlanFragment` here ?

arhimondr replied:
`PlanFragment` is not a tree node. `SubPlan` is. And it looks like linking together `PlanFragment`'s into a tree is the only responsibility of the `SubPlan` class. `PlanFragment` defines relations with other fragments already, see: `PlanFragment#getRemoteSourceNodes`. The `RemoteSourceNode` contains the `List<PlanFragmentId> sourceFragmentIds`.

wenleix commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -593,22 +705,29 @@ else if (!result.getBlocked().isDone()) {
                 }
             }
+
+            scheduling.set(false);
```
So, `scheduling` variable guard only one thread can enter the previous code section.

It's not immediately clear to me what happens, if one thread is scheduling stages, while another thread is trying to get the next batch of stages to be scheduled.

arhimondr replied:
> It's not immediately clear to me what happens, if one thread is scheduling stages, while another thread is trying to get the next batch of stages to be scheduled.

The second thread will check the `scheduling` variable, and if it is set to `true` - it won't start scheduling. Whenever scheduling is finished - the `schedule` method set's the flag back to false, and checks when there's anything to schedule.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/DistributedExecutionPlanner.java:
```diff
@@ -144,11 +145,13 @@ private Visitor(Session session, StageExecutionDescriptor stageExecutionDescript
                     node.getLayout().get(),
                     stageExecutionDescriptor.isScanGroupedExecution(node.getId()) ? GROUPED_SCHEDULING : UNGROUPED_SCHEDULING);
 
+            SplitSource splitSource = node.isTemporaryTable() ? new LazySplitSource(splitSourceSupplier::get) : splitSourceSupplier.get();
```
nit: pass in `splitSourceSupplier` is enough. No `::get` needed :)

tdcmeehan replied:
It won't work because here we use Guava and in LazySplitSource we use java.util

tdcmeehan replied:
I think we can switch here to use java.util Supplier

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LazySplitSource.java:
```diff
@@ -0,0 +1,94 @@
+    private final Supplier<SplitSource> supplier;
+
+    @GuardedBy("this")
+    private SplitSource instance;
```
why not calling this delegate?

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LazySplitSource.java:
```diff
@@ -0,0 +1,94 @@
+        closed = true;
+    }
+
+    public synchronized SplitSource getDelegate()
```
private?

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LazySplitSource.java:
```diff
@@ -0,0 +1,94 @@
+    public synchronized void close()
+    {
+        // not yet initialized
+        if (instance == null) {
```
Can `close` be called before it's initialized ?

arhimondr replied:
Yeah, in theory.

wenleix commented on presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java:
```diff
@@ -141,6 +153,20 @@ public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> ta
+                        "query.materialize-exchanges-strategy", "ALL",
+                        "query.hash-partition-count", "5",
+                        "colocated-joins-enabled", "true",
+                        "grouped-execution-for-aggregation-enabled", "true"),
```
Also set concurrent lifespan per task to be 1 :)

arhimondr replied:
It is `1` by default. Do we want to have it set so explicitly?

wenleix commented on presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java:
```diff
@@ -141,6 +153,20 @@ public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> ta
+                ImmutableMap.of(
+                        "query.partitioning-provider-catalog", "hive",
+                        "query.materialize-exchanges-strategy", "ALL",
+                        "query.hash-partition-count", "5",
```
why not the classic number of `11`? This has more rounds for grouped execution :)

arhimondr replied:
No reason. Let me change that.

wenleix commented:
I make an initial pass, generally looks on the right direction. And the major comments are (not surprisingly) in the following two commits:

- "Support materialized exchanges in PlanFragmenter"
- "Support dependent stages in SqlQueryScheduler"

One meta-question is, given the size of the PR, do we want to split it into "planner" part and "execution" part? Each part will include exact one "monster" commit. 

At the end of the "planner" part PR, it will include a very preliminary plan assertion test (1 simple join query and 1 simple aggregation query would be enough) for sanity check.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java:
```diff
@@ -43,6 +45,7 @@
 
     private int hashPartitionCount = 100;
     private String partitioningProviderCatalog = GlobalSystemConnector.NAME;
+    private MaterializeExchangesStrategy materializeExchangesStrategy = NONE;
```
In this case

> private **String** partitioningProviderCatalog = **GlobalSystemConnector**.NAME;

The `GlobalSystemConnector` prefix is needed to disambiguate what connector `NAME` do we set. As the field itself is a `String`. And the `NAME` itself doesn't carry any useful information. Similarly to ImmutableList.**of()**

However in

> **MaterializeExchangesStrategy** materializeExchangesStrategy = NONE;

It is obvious that the assigned type is `MaterializeExchangesStrategy`, and it is set to `NONE` which is self descriptive.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -1251,6 +1258,15 @@ private Partitioning createPartitioning(List<Symbol> partitioningColumns)
+
+        private Scope getExchangeScope(PlanNode input)
+        {
+            if (input.getOutputSymbols().isEmpty()) {
```
There are some corner cases related to `JOIN` + `count(*)`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -389,6 +433,242 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+            checkArgument(exchange.getType() == REPARTITION, "Unexpected exchange type: %s", exchange.getType());
+
+            PartitioningScheme partitioningScheme = exchange.getPartitioningScheme();
+            checkArgument(!partitioningScheme.getHashColumn().isPresent(), "precomputed hashes are not supported in materializing exchanges");
```
Yes. Currently we support precomputed hashes in exhange only for system partitioning.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -389,6 +433,242 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+            Partitioning partitioning = partitioningScheme.getPartitioning();
+            List<Symbol> partitioningSymbols = assignPartitioningSymbols(partitioning, constantExpressions);
+            Map<Symbol, ColumnMetadata> temporaryTableColumns = assignColumnNames(exchange.getOutputSymbols(), constantExpressions.keySet());
+            List<String> partitioningColumns = partitioningSymbols.stream()
```
I renamed to `partitionColumns` and `partitionSymbols`. Do you think they should go singular?

arhimondr commented:
@wenleix 

> Generally looks good. I have a question about whether we want to pass the "parentNode" to getExchangeScope. -- Otherwise, later it might be hard to understand why some remote partitioned exchange will call getExchangeScope, while some are hard coded to REMOTE.

Discussed offline. Hardcoded `REMOTE` are for pre-bucketed cases.

arhimondr commented:
@wenleix @rschlussel @tdcmeehan Addressed comments and extracted the first part (planner) into a separate PR: https://github.com/prestodb/presto/pull/12568

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -216,6 +226,43 @@ private SqlQueryScheduler(
+        Set<PlanFragmentId> rootIds = new HashSet<>();
+        rootIds.add(plan.getFragment().getId());
+        // dependencies are roots
+        plan.geAllStages()
```
Is `ge` short for something or is this a typo?

arhimondr replied:
That's a typo. Let me fix it

arhimondr replied:
This code has been removed

tdcmeehan commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveMaterializedQueries.java:
```diff
@@ -88,6 +92,102 @@ public void testValues()
+    }
+
+    @Override
+    public void testMergeHyperLogLogGroupBy()
```
Just checking my understanding, is it because we are materializing to ORC format and there is no type in Hive for HLL?  If so I suppose with "raw" format we could enable them?

arhimondr replied:
Unfortunately Hive cannot serialize HLL type into a table. That's why these queries are not yet supported.

wenleix commented:
@arhimondr 

> BTW: I think UI is currently broken when you materialize exchange.... I think the trick is to fix SubPlan#getAllFragments

> Since we no longer have 2 lists, i don't think it should be needeed

Ah, I am wrong, it might be `SqlStageScheduler::buildStageInfo`

arhimondr commented:
@wenleix 

> I believe isFinishedForLifespan is in theory not implemented correctly

Could you please elaborate a little bit more on that?

> I understand we don't want to "officially" introduce "PlanSection" for now due to the complication. 

I agree that having these maps is confusing. I introduced a private class `StreamingPlanSection`. With `StreamingPlanSection` it is possible to build a forest of plan fragments connected with the streaming exchange. I believe it simplifies the code and also is a good step toward retriability direction.

> If one thread is scheduling stages, while another thread is trying to get the next batch of stages to be scheduled, it's not clear to me what will happen.

The logic is next:

1. Only a single scheduling can be happening any moment of time. It is guaranteed by the atomic `scheduling` property.
2. Whenever a stage is finished - we check two conditions. If there are any new stages that are ready to be scheduled. If there are - we are trying to schedule them if no other scheduling is in process. 
3. Once scheduling is finished - we check if there any more stages to be scheduled (ones that we didn't start scheduling on step 2)

> We can also start with only run one "plan section" at a time.

That's the behaviour right now if the phased execution is enabled. For all-at-once it is all at once.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -401,10 +454,15 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
         }
 
-        ImmutableSet.Builder<SqlStageExecution> childStagesBuilder = ImmutableSet.builder();
+        ImmutableSet.Builder<SqlStageExecution> inputStagesBuilder = ImmutableSet.builder();
```
After introduction of `StreamingPlanSection` this code is no longer needed. Now the `createStages` method schedules streaming connected trees in the forest.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StageExecutionPlan.java:
```diff
@@ -73,6 +82,25 @@ public StageExecutionPlan withBucketToPartition(Optional<int[]> bucketToPartitio
         return new StageExecutionPlan(fragment.withBucketToPartition(bucketToPartition), splitSources, subStages);
     }
 
+    public Stream<StageExecutionPlan> geAllStages()
```
This method has been removed and replaced with the `StreamingPlanSection` abstraction. Also the `StageExecutionPlan` has been removed.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -216,6 +226,43 @@ private SqlQueryScheduler(
+
+        Set<PlanFragmentId> rootIds = new HashSet<>();
+        rootIds.add(plan.getFragment().getId());
+        // dependencies are roots
```
no longer actual

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -216,6 +226,43 @@ private SqlQueryScheduler(
+        this.remoteSourceLinkedSubtrees = remoteSourceLinkedSubtrees.build();
+
+        // collect dependencies per remote source linked subtree
+        ImmutableMap.Builder<PlanFragmentId, Set<PlanFragmentId>> remoteSourceLinkedSubtreeDependencies = ImmutableMap.builder();
```
no longer actual. Replaced with the `StreamingPlanSection` forest

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -216,6 +226,43 @@ private SqlQueryScheduler(
+                .forEach(rootIds::add);
+
+        List<StageExecutionPlan> remoteSourceLinkedSubtreeRoots = stream(forTree(StageExecutionPlan::getSubStages).depthFirstPreOrder(plan))
+                .filter(node -> rootIds.contains(node.getFragment().getId()))
```
Ditto

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -216,6 +226,43 @@ private SqlQueryScheduler(
+        this.stageByFragmentIdMap = stages.stream()
+                .collect(toImmutableMap(stage -> stage.getFragment().getId(), identity()));
+
+        Set<PlanFragmentId> rootIds = new HashSet<>();
```
No longer actual

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -525,15 +596,56 @@ public Duration getTotalCpuTime()
+        for (Entry<PlanFragmentId, List<PlanFragmentId>> remoteSourceLinkedSubtree : remoteSourceLinkedSubtrees.entrySet()) {
+            PlanFragmentId subtreeRootId = remoteSourceLinkedSubtree.getKey();
+            Set<PlanFragmentId> subtreeDependencies = remoteSourceLinkedSubtreeDependencies.get(subtreeRootId);
+            if (!allStagesFinished(subtreeDependencies)) {
```
Ditto

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -249,6 +296,12 @@ else if (state == ABORTED) {
                     // this should never happen, since abort can only be triggered in query clean up after the query is finished
                     queryStateMachine.transitionToFailed(new PrestoException(GENERIC_INTERNAL_ERROR, "Query stage was aborted"));
                 }
+                else if (state.isDone() && !state.isFailure()) {
```
Changed to `state == FINISHED`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -249,6 +296,12 @@ else if (state == ABORTED) {
                     queryStateMachine.transitionToFailed(new PrestoException(GENERIC_INTERNAL_ERROR, "Query stage was aborted"));
                 }
+                else if (state.isDone() && !state.isFailure()) {
+                    List<SqlStageExecution> nextStagesToSchedule = selectStagesWithFinishedDependencies();
```
https://github.com/prestodb/presto/pull/12469/commits/b6e38a665931801a57d5bcf41268a1d40fa09c05#diff-5cffc1bea86a011bf263a6d6a736c93fR615

We add only `PLANNED` stages to scheduling. I changed the condition a bit.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -422,20 +480,33 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
 
-            SqlStageExecution childStage = subTree.get(0);
-            childStagesBuilder.add(childStage);
+            SqlStageExecution root = subTree.get(0);
```
I inlined it

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -114,13 +120,17 @@
     private final QueryStateMachine queryStateMachine;
     private final ExecutionPolicy executionPolicy;
     private final Map<StageId, SqlStageExecution> stages;
```
Now i only have a single map, `planFragmentIdToStageIdMap`. I left the `stages` map untouched.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StageExecutionPlan.java:
```diff
@@ -73,6 +82,25 @@ public StageExecutionPlan withBucketToPartition(Optional<int[]> bucketToPartitio
         return new StageExecutionPlan(fragment.withBucketToPartition(bucketToPartition), splitSources, subStages);
     }
 
+    public Stream<StageExecutionPlan> geAllStages()
```
This code has been removed

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StageExecutionPlan.java:
```diff
@@ -73,6 +82,25 @@ public StageExecutionPlan withBucketToPartition(Optional<int[]> bucketToPartitio
+        return stream(forTree(StageExecutionPlan::getSubStages).depthFirstPreOrder(this));
+    }
+
+    public Stream<StageExecutionPlan> geAllRemoteSourceLinkedStages()
```
ditto

arhimondr commented:
Superseeded by: https://github.com/prestodb/presto/pull/12604
